rmw_fastrtps_shared_cpp/src/participant.cpp

```diff
@@ -0,0 +1,265 @@
+      allocator.deallocate(file_path, allocator.state);
+      return false;
+    }
+
+    allocator.deallocate(file_path, allocator.state);
+  }
+
+  return true;
+}
+#endif
+
+static
+CustomParticipantInfo *
+__create_participant(
+  const char * identifier,
+  ParticipantAttributes participantAttrs,
+  bool leave_middleware_default_qos,
+  rmw_dds_common::Context * common_context)
+{
+  // Declare everything before beginning to create things.
+  ::ParticipantListener * listener = nullptr;
+  Participant * participant = nullptr;
+  CustomParticipantInfo * participant_info = nullptr;
+
+  try {
+    listener = new ::ParticipantListener(
+      identifier, common_context);
+  } catch (std::bad_alloc &) {
+    RMW_SET_ERROR_MSG("failed to allocate participant listener");
+    goto fail;
+  }
+
+  participant = Domain::createParticipant(participantAttrs, listener);
+  if (!participant) {
+    RMW_SET_ERROR_MSG("create_node() could not create participant");
+    return nullptr;
+  }
+
+  try {
+    participant_info = new CustomParticipantInfo();
+  } catch (std::bad_alloc &) {
+    RMW_SET_ERROR_MSG("failed to allocate node impl struct");
+    goto fail;
+  }
+  participant_info->leave_middleware_default_qos = leave_middleware_default_qos;
+
+  participant_info->participant = participant;
+  participant_info->listener = listener;
+
+  return participant_info;
+fail:
+  rmw_free(listener);
+  if (participant) {
+    Domain::removeParticipant(participant);
+  }
+  return nullptr;
+}
+
+CustomParticipantInfo *
+rmw_fastrtps_shared_cpp::create_participant(
+  const char * identifier,
+  size_t domain_id,
+  const rmw_security_options_t * security_options,
+  bool localhost_only,
+  const char * security_context,
+  rmw_dds_common::Context * common_context)
+{
+  if (!security_options) {
+    RMW_SET_ERROR_MSG("security_options is null");
+    return nullptr;
+  }
+  ParticipantAttributes participantAttrs;
+
+  // Load default XML profile.
+  Domain::getDefaultParticipantAttributes(participantAttrs);
+
+  participantAttrs.rtps.builtin.domainId = static_cast<uint32_t>(domain_id);
+
+  if (localhost_only) {
+    Locator_t local_network_interface_locator;
+    static const std::string local_ip_name("127.0.0.1");
+    local_network_interface_locator.kind = 1;
+    local_network_interface_locator.port = 0;
+    IPLocator::setIPv4(local_network_interface_locator, local_ip_name);
+    participantAttrs.rtps.builtin.metatrafficUnicastLocatorList.push_back(
+      local_network_interface_locator);
+    participantAttrs.rtps.builtin.initialPeersList.push_back(local_network_interface_locator);
+  }
+
+  participantAttrs.rtps.builtin.domainId = static_cast<uint32_t>(domain_id);
+
+  size_t length = strlen(security_context) + strlen("securitycontext=;") + 1;
+  participantAttrs.rtps.userData.resize(length);
+  int written = snprintf(
+    reinterpret_cast<char *>(participantAttrs.rtps.userData.data()),
+    length, "securitycontext=%s;", security_context);
```

Comment:
@ivanpauno nit^2: calling `snprintf` with a `null` buffer of zero `length` will get you the required size, avoiding the need for clever computations.

Reply:
Nice trick, I checked cppreference to confirm https://en.cppreference.com/w/cpp/io/c/fprintf.
See https://github.com/ros2/rmw_fastrtps/pull/312/commits/f043aaa8eeec2471fa7a16e9dc3f143d77eca0ff.

Reply:
See also https://github.com/ros2/rmw_fastrtps/pull/312/commits/fa959e1d8c387e657b127e5ae93de5eac4202a31.